xarray/core/indexing.py

```diff
@@ -296,15 +296,16 @@ def slice_slice(old_slice: slice, applied_slice: slice, size: int) -> slice:
 
 
 def _index_indexer_1d(old_indexer, applied_indexer, size: int):
-    assert isinstance(applied_indexer, integer_types + (slice, np.ndarray))
     if isinstance(applied_indexer, slice) and applied_indexer == slice(None):
         # shortcut for the usual case
         return old_indexer
     if isinstance(old_indexer, slice):
         if isinstance(applied_indexer, slice):
             indexer = slice_slice(old_indexer, applied_indexer, size)
+        elif isinstance(applied_indexer, integer_types):
+            indexer = range(*old_indexer.indices(size))[applied_indexer]  # type: ignore[assignment]
```

Comment:
These kinds of `# type: ignore[assignment]` errors usually means `indexer` has changed type, which usually leads to requiring another costly `isinstance` again somewhere downstream.

Reply:
This one is due to the ambiguity between slice and int for these indexers. I can look into it if you want.

However. It starts to touch some tuple to lists conversation that was identified as something I shouldn't adjust too much due to potential conflicts with an active in development branch.

Reply:
<details><summary>The following patch would</summary>

```
diff --git a/xarray/core/indexing.py b/xarray/core/indexing.py
index e29d77da..44560dbd 100644
--- a/xarray/core/indexing.py
+++ b/xarray/core/indexing.py
@@ -9,7 +9,7 @@ from contextlib import suppress
 from dataclasses import dataclass, field
 from datetime import timedelta
 from html import escape
-from typing import TYPE_CHECKING, Any, Callable, overload
+from typing import TYPE_CHECKING, Any, Callable, overload, List

 import numpy as np
 import pandas as pd
@@ -295,20 +295,24 @@ def slice_slice(old_slice: slice, applied_slice: slice, size: int) -> slice:
     return slice(start, stop, step)


-def _index_indexer_1d(old_indexer, applied_indexer, size: int):
+def _index_indexer_1d(
+    old_indexer: slice | np.ndarray[Any, np.dtype[np.generic]],
+    applied_indexer: slice | int | np.integer | np.ndarray[Any, np.dtype[np.generic]],
+    size: int
+) -> slice | int | np.integer | np.ndarray[Any, np.dtype[np.generic]]:
+    """Helper routine to slice non integers into an entry for OuterIndexer"""
     if isinstance(applied_indexer, slice) and applied_indexer == slice(None):
         # shortcut for the usual case
         return old_indexer
     if isinstance(old_indexer, slice):
         if isinstance(applied_indexer, slice):
-            indexer = slice_slice(old_indexer, applied_indexer, size)
+            return slice_slice(old_indexer, applied_indexer, size)
         elif isinstance(applied_indexer, integer_types):
-            indexer = range(*old_indexer.indices(size))[applied_indexer]  # type: ignore[assignment]
+            return range(*old_indexer.indices(size))[applied_indexer]
         else:
-            indexer = _expand_slice(old_indexer, size)[applied_indexer]
+            return _expand_slice(old_indexer, size)[applied_indexer]
     else:
-        indexer = old_indexer[applied_indexer]
-    return indexer
+        return old_indexer[applied_indexer]


 class ExplicitIndexer:
@@ -630,12 +634,14 @@ class LazilyIndexedArray(ExplicitlyIndexedNDArrayMixin):

     def _updated_key(self, new_key: ExplicitIndexer) -> BasicIndexer | OuterIndexer:
         iter_new_key = iter(expanded_indexer(new_key.tuple, self.ndim))
-        full_key = []
+        full_key : List[slice | int | np.integer | np.ndarray[Any, np.dtype[np.generic]]] = []
+        outer_indexer = True
         for size, k in zip(self.array.shape, self.key.tuple):
             if isinstance(k, integer_types):
                 full_key.append(k)
             else:
                 full_key.append(_index_indexer_1d(k, next(iter_new_key), size))
+
         full_key_tuple = tuple(full_key)

         if all(isinstance(k, integer_types + (slice,)) for k in full_key_tuple):
```
</details>

would move the error to
```
xarray/core/indexing.py:648: error: Argument 1 to "BasicIndexer" has incompatible type "tuple[slice | int | integer[Any] | ndarray[Any, dtype[generic]], ...]"; expected "tuple[int | integer[Any] | slice, ...]"  [arg-type]
```

While I don't disagree that:

> which usually leads to requiring another costly isinstance again somewhere downstream.

I would rather keep this for a future change once the branch `backend-indexing` has been merge in.

Reply:
Sorry one more request here. Can we just update `slice_slice` with the faster method, assuming its faster?